core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -2047,8 +2047,15 @@ protected void convertFrom(
 
     case AS:
       call = (SqlCall) from;
-      convertFrom(bb, call.operand(0));
-      if (call.operandCount() > 2
+      SqlNode firstOperand = call.operand(0);
+
+      if (firstOperand.getKind() == SqlKind.UNNEST) {
+        convertUnnest(bb, (SqlCall) firstOperand, getFieldAliases(call));
+      } else {
+        convertFrom(bb, firstOperand);
+      }
```

Comment:
The operator row type belongs to the Validator namespace, i think we should fix it there instead of `SqlToRelConverter`.

Reply:
Yea, I considered it. For example, for query
{code}
select d.deptno, e, k.empno\n"
        + "from dept_nested_expanded as d CROSS JOIN\n"
        + " UNNEST(d.admins, d.employees) as t(e, k)
{code}
After validation, it will become
{code}
select d.deptno, t.e, t.k.empno\n"
        + "from dept_nested_expanded as d CROSS JOIN\n"
        + " UNNEST(d.admins, d.employees) as t(e, k)
{code}
The failure comes from the Project. Since the rowtype of Uncollect is already flattened, so Calcite could not find the path t.e and t.k.empo from the input, which is the Uncollect RelNode.

So my change will keep them nested, as in SqlUnnestOperator, and I think it's also aligned with the expected behavior, which is unnesting array but not unbox the component struct. 

Could you elaborate on your thoughts on AliasNamespace? And if possible, could you provide an ideal logical plan looks like? Thanks!

Reply:
For type flattern, fix the `RelStructuredTypeFlattener` for the operator you care about instead of the `SqlToRelConverter`.

Reply:
@danny0405 I don't quite your point. If you are talking about the decorrelate, I will fix it. But this part of change is not about decorrelate. Could you elaborate more on how does your fix work in more detail? If it's a fundamental change, could you please post it in the ticket too? I will also describe my plan there. Thanks.

Reply:
I agree, AliasNamespace sounds like a good direction to explore.

Reply:
@danny0405 @julianhyde I revisited the flow. I will use UNNEST(d.employees) as t(k) as an example.

In validation phase, since this is a AS call and Calcite produces an AliasNamespace. And since we disable flatten of STRUCT type for SqlUnnestOperator, the output type is "expected", with one field `k` and its type is employee struct type.

During SqlToRel, the problem is in bb.lookupExp in convertIdentifer (e.g. T.K.EMPO). Basically the identifier could be resolved from namespace, the type is also expected. When `lookup` RexNode in inputs, the type output is wrong because the input type is already flattened by Uncollect:deriveRowType. 

Basically, we either modify Uncollect.deriveRowType to produce the properly aliased unflattend row type, or we modify the `lookup` Rexnode logic, and maybe some other places. The second one is pretty de-centralized so I suggest to take the first one.

For the first solution, my implementation is one option. Another option is to directly use the right namespace's rowType in Uncollect.deriveRowType. One question is that I see generally RelNode type deriving is separate from validation results (namespace's rowType). May I know why? And do you think it's gonna work for this use case? 

Thanks.

Reply:
Did you notice that i have given a fix patch on the issue ? [1]

[1] https://issues.apache.org/jira/browse/CALCITE-3789

Reply:
@danny0405 Sorry was a little busy recently and just noticed it. Added one nit commit. Your way to plug in the aliases is really clean 👍

Reply:
@danny0405 patch greatly improves the first approach. Resolve this comment.